rfcs/0000-oidc-policy.md

```diff
@@ -0,0 +1,183 @@
+# `OpenIDConnectPolicy` - a metapolicy
+
+- Feature Name: oidc-policy
+- Start Date: 2025-02-05
+- RFC PR: [Kuadrant/architecture#0114](https://github.com/Kuadrant/architecture/pull/114)
+- Issue tracking: [Kuadrant/architecture#0000](https://github.com/Kuadrant/architecture/issues/0000)
+
+# Summary
+[summary]: #summary
+
+The `OpenIDConnectPolicy` aims at helping users to easily & quickly get started with the Kuadrant stack to setup an
+[OpenID Connect flow](https://openid.net/developers/how-connect-works/) without needing to initially understand how to
+use Kuadrant's `AuthPolicy` to achieve their goals. 
+
+# Motivation
+[motivation]: #motivation
+
+There are a few goals this implementation hopes to achieve and set a precedent for introducing more such APIs in
+Kuadrant:
+
+ - users looking at an example policy would be able to make mostly sense of it without needing to lookup documentation;
+ - the policy would apply sensible defaults in most cases and only require the very minimal input from the user;
+ - the idea is to implement it as _metapolicy_, which would only be consumed by a controller to produce a `AuthPolicy`
+   that would then enforce the flow;
+ - it is _not_ the goal to expose all the power of the `AuthPolicy` to the user through the new policy, but rather to let
+   users achieve most of their goals by slowly exploring defaults of the policy and eventually dig down the stack, into
+   the `AuthPolicy` to expand their use cases when they need it;
+ - explore how to, overtime, grow the initial minimal API of the policy into a larger one to cover more use cases than
+   the one currently set as the goal for this RFC
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `OpenIDConnectPolicy` is a direct policy attachment that attaches to one or multiple
+[`HttpRoute`](https://gateway-api.sigs.k8s.io/api-types/httproute/) objects. Targeted routes will require requests to
+carry an access token to be authorized through. Should the token be missing or be invalid, the requesting entity gets
+redirected to the `provider`'s `authorizationEndpoint` specified in the policy for them to identify.
+
+It is important to note here that the access token is to be negotiated stored by the client. While this may change in
+the future, currently the token is to either be provided by the appropriate HTTP header (`Authorization: Bearer`, which 
+is the default source), or by a Cookie.
+
+## Minimal example
+
+Using a very minimal example, we'll use
+[gitlab.com](https://docs.gitlab.com/ee/integration/openid_connect_provider.html) as our provider for an OpenID Connect
+flow to protect all requests routed to our application by the `HttpRoute`.
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: OpenIDConnectPolicy
+metadata:
+  name: gitlab
+spec:
+  targetRef:
+    name: toystore
+    group: gateway.networking.k8s.io
+    kind: HttpRoute
+  provider:
+    host: gitlab.com
```

Comment:
I'm guessing `host` here would be the issuer URL?

Reply:
Maybe name this field `issuerUrl`?

Reply:
`issuerHost`? Or do we require a URL then?, i.e. `https://gitlab.com`? Or is it a URI even then?

Reply:
maybe `provider` was wrong to begin with, or does look like it according to [this](https://openid.net/specs/openid-connect-core-1_0.html#Terminology)

Reply:
I'm afraid host is not enough. The not well-known part in Keycloak for example includes a path prefix always. Then you can append the `/.well-known/openid-configuration` bit to it.

Reply:
This can be addressed by specifying the `discoveryEndpoint` property in that case, or am I missing something?

Reply:
Not saying it shouldn't be changed, but here is my understanding: there is no "issuer endpoint", there is authorization, token et al endpoints. So my thinking was that `host` is just a shorthand, that's generally understood and all that's needed _and isn't mandatory_. So if you want to specify _any_ OpenID Connect (Discovery) endpoint, they would exists and be an option _always_.

Reply:
https://www.rfc-editor.org/rfc/rfc9207.html

Reply:
Also, if we did that (which I'm cool with btw), would we then revert back the parent from `issuer` to `provider`? So that:

```yaml
provider:
  issuer: https://gitlab.com
```

Reply:
> so are you arguing to make `host` -> `issuer`?

Yes. Unless I misunderstood it and the `host` is not the thing that identifies and lets us find the auth server.

> would we then revert back the parent from `issuer` to `provider`?

Maybe I'm missing something, but the parent says `provider` now in the text.

Anyway, I would keep `provider` as is, rename the `host` field as `issuer`, and make it a URL instead of just a host name－exactly as you suggested I guess, i.e.:

```yaml
provider:
  issuer: https://gitlab.com
```

Reply:
Yeah, I hadn't pushed the latest changes to not loose these discussions. Dunno where/how it was decided but committed that change yesterday locally. Let me revert and use `issuer` as a URL then!